In [41]:
!pip install pandas
!pip install pandas_datareader
!pip install yfinance

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 109.5/109.5 kB 1.9 MB/s eta 0:00:00MB/s eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.0/5.0 MB 5.5 MB/s eta 0:00:00m eta 0:00:010:00:01
  Using cached appdirs-1.4.4-py2.py3-none-any.whl.metadata (9.0 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 315.3/315.3 kB 791.9 kB/s eta 0:00:000:00:010:00:01:01
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 2.1 MB/s eta 0:00:00m eta 0:00:010:01:01
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.0/73.0 kB 2.2 MB/s eta 0:00:00
Using cached appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 112.2/112.2 kB 2.4 MB/s eta 0:00:003.1 MB/s eta 0:00

In [1]:
import pandas as pd

### Question 1: [Macro] Average growth of GDP in 2023

**What is the average growth (in %) of GDP in 2023?**

Download the timeseries Real Gross Domestic Product (GDPC1) from FRED (https://fred.stlouisfed.org/series/GDPC1). 
Calculate year-over-year (YoY) growth rate (that is, divide current value to one 4 quarters ago). Find the average YoY growth in 2023 (average from 4 YoY numbers).
Round to 1 digit after the decimal point: e.g. if you get 5.66% growth => you should answer  5.7

In [4]:
# Load the dataset
df = pd.read_csv('../../03-dataset/01-fred/GDPC1-01-2022-10-2023.csv')


In [5]:
# Convert the date column to datetime
df['DATE'] = pd.to_datetime(df['DATE'])

# Set the date column as the index
df.set_index('DATE', inplace=True)

In [6]:
# Calculate year-over-year growth rates
df['YoY_Growth'] = df['GDPC1'].pct_change(periods=4)

In [11]:
num_rows, num_cols = df.shape
print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_cols}")

df.head(num_rows)

Number of rows: 8
Number of columns: 2


GDPC1  YoY_Growth
DATE                             
2022-01-01  21738.871         NaN
2022-04-01  21708.160         NaN
2022-07-01  21851.134         NaN
2022-10-01  21989.981         NaN
2023-01-01  22112.329    0.017179
2023-04-01  22225.350    0.023825
2023-07-01  22490.692    0.029269
2023-10-01  22679.255    0.031345

In [7]:
# Filter for 2023 data
df_2023 = df[df.index.year == 2023]

# Calculate average YoY growth for 2023
avg_yoy_2023 = df_2023['YoY_Growth'].mean() * 100


The average year-over-year GDP growth rate for 2023 is: 2.5%


In [13]:
print(avg_yoy_2023)

2.5404433928676253


In [14]:
# Round to 1 decimal place
avg_yoy_2023_rounded = round(avg_yoy_2023, 1)

print(f"The average year-over-year GDP growth rate for 2023 is: {avg_yoy_2023_rounded}%")

The average year-over-year GDP growth rate for 2023 is: 2.5%


---

### Question 2. [Macro] Inverse "Treasury Yield"

**Find the min value of (dgs10-dgs2) after since year 2000 (2000-01-01) and write it down as an answer, round to 1 digit after the decimal point.**


Download DGS2 and DGS10 interest rates series (https://fred.stlouisfed.org/series/DGS2,
 https://fred.stlouisfed.org/series/DGS10). Join them together to one dataframe on date (you might need to read about pandas.DataFrame.join()), calculate the difference dgs10-dgs2 daily.

(Additional: think about what does the "inverted yield curve" mean for the market and investors? do you see the same thing in your country/market of interest? Do you think it can be a good predictive feature for the models?)

In [28]:
# Load the dataset
df_dgs2_bz = pd.read_csv('../../03-dataset/01-fred/DGS2-01-2000-04-2024.csv')
df_dgs10_bz = pd.read_csv('../../03-dataset/01-fred/DGS10-01-2000-04-2024.csv')


In [29]:
num_rows, num_cols = df_dgs2_bz.shape
print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_cols}")

df_dgs2_bz.head(5)


Number of rows: 6342
Number of columns: 2


DATE   DGS2
0  2000-01-03  6.380
1  2000-01-04  6.300
2  2000-01-05  6.380
3  2000-01-06  6.350
4  2000-01-07  6.310

In [30]:
num_rows, num_cols = df_dgs10_bz.shape
print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_cols}")

df_dgs10_bz.head(5)

Number of rows: 6342
Number of columns: 2


DATE DGS10
0  2000-01-03  6.58
1  2000-01-04  6.49
2  2000-01-05  6.62
3  2000-01-06  6.57
4  2000-01-07  6.52

In [31]:
# Convert 'DGS2' and 'DGS10' columns to float, coercing invalid values to NaN
df_dgs2_bz['DGS2'] = pd.to_numeric(df_dgs2_bz['DGS2'], errors='coerce')
df_dgs10_bz['DGS10'] = pd.to_numeric(df_dgs10_bz['DGS10'], errors='coerce')

# Clean up missing values if necessary
df_dgs2_bz.dropna(subset=['DGS2'], inplace=True)
df_dgs10_bz.dropna(subset=['DGS10'], inplace=True)


In [32]:
# Convert the date column to datetime
df_dgs2_bz['DATE'] = pd.to_datetime(df_dgs2_bz['DATE'])
df_dgs10_bz['DATE'] = pd.to_datetime(df_dgs10_bz['DATE'])

# Set the date column as the index
df_dgs2_bz.set_index('DATE', inplace=True)
df_dgs10_bz.set_index('DATE', inplace=True)

In [33]:
num_rows, num_cols = df_dgs2_bz.shape
print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_cols}")

df_dgs2_bz.head(5)


Number of rows: 6081
Number of columns: 1


DGS2
DATE            
2000-01-03  6.38
2000-01-04  6.30
2000-01-05  6.38
2000-01-06  6.35
2000-01-07  6.31

In [34]:
num_rows, num_cols = df_dgs10_bz.shape
print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_cols}")

df_dgs10_bz.head(5)

Number of rows: 6081
Number of columns: 1


DGS10
DATE             
2000-01-03   6.58
2000-01-04   6.49
2000-01-05   6.62
2000-01-06   6.57
2000-01-07   6.52

In [38]:
# Join the dataframes on the 'DATE' column
df_dgs_joined = df_dgs2_bz.join(df_dgs10_bz, how='inner')

# Calculate the difference between 'DGS10' and 'DGS2'
df_dgs_joined['DIFF'] = df_dgs_joined['DGS10'] - df_dgs_joined['DGS2']

# Reset the index to make 'DATE' a regular column again
df_dgs_joined.reset_index(inplace=True)
df_dgs_joined.set_index('DATE', inplace=True)

# Print the first few rows of the combined dataframe
print(df_dgs_joined.head())

            DGS2  DGS10  DIFF
DATE                         
2000-01-03  6.38   6.58  0.20
2000-01-04  6.30   6.49  0.19
2000-01-05  6.38   6.62  0.24
2000-01-06  6.35   6.57  0.22
2000-01-07  6.31   6.52  0.21


In [39]:
num_rows, num_cols = df_dgs_joined.shape
print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_cols}")

df_dgs_joined.head(5)

Number of rows: 6081
Number of columns: 3


DGS2  DGS10  DIFF
DATE                         
2000-01-03  6.38   6.58  0.20
2000-01-04  6.30   6.49  0.19
2000-01-05  6.38   6.62  0.24
2000-01-06  6.35   6.57  0.22
2000-01-07  6.31   6.52  0.21

In [40]:
min_diff = round(df_dgs_joined['DIFF'].min(), 1)
print(f"The minimum value of DIFF is: {min_diff}")

The minimum value of DIFF is: -1.1


---
### Question 3. [Index] Which Index is better recently?

**Compare S&P 500 and IPC Mexico indexes by the 5 year growth and write down the largest value as an answer (%)**

Download on Yahoo Finance two daily index prices for S&P 500 (^GSPC, https://finance.yahoo.com/quote/%5EGSPC/) and IPC Mexico (^MXX, https://finance.yahoo.com/quote/%5EMXX/). Compare 5Y growth for both (between 2019-04-09 and 2024-04-09). Select the higher growing index and write down the growth in % (closest integer %). E.g. if ratio end/start was 2.0925 (or growth of 109.25%), you need to write down 109 as your answer.

(Additional: think of other indexes and try to download stats and compare the growth? Do create 10Y and 20Y growth stats. What is an average yearly growth rate (CAGR) for each of the indexes you select?)


In [42]:
from pandas_datareader import data as pdr
import yfinance as yf
yf.pdr_override()

In [44]:
# Set the start and end dates:
start_date = '2019-04-09'
end_date = '2024-04-09'

# Download the data for S&P 500 and IPC Mexico:
df_sp500_bz = pdr.get_data_yahoo('^GSPC', start=start_date, end=end_date)
df_ipc_mexico_bz = pdr.get_data_yahoo('^MXX', start=start_date, end=end_date)



[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed


In [45]:
num_rows, num_cols = df_sp500_bz.shape
print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_cols}")

df_sp500_bz.head(5)

Number of rows: 1258
Number of columns: 6


Open         High          Low        Close    Adj Close  \
Date                                                                          
2019-04-09  2886.580078  2886.879883  2873.330078  2878.199951  2878.199951   
2019-04-10  2881.370117  2889.709961  2879.129883  2888.209961  2888.209961   
2019-04-11  2891.919922  2893.419922  2881.989990  2888.320068  2888.320068   
2019-04-12  2900.860107  2910.540039  2898.370117  2907.409912  2907.409912   
2019-04-15  2908.320068  2909.600098  2896.479980  2905.580078  2905.580078   

                Volume  
Date                    
2019-04-09  3032480000  
2019-04-10  3092230000  
2019-04-11  2970650000  
2019-04-12  3726050000  
2019-04-15  3114530000

In [46]:
num_rows, num_cols = df_ipc_mexico_bz.shape
print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_cols}")

df_ipc_mexico_bz.head(5)

Number of rows: 1258
Number of columns: 6


Open          High           Low         Close  \
Date                                                                 
2019-04-09  45346.828125  45346.828125  44854.531250  45151.628906   
2019-04-10  45204.750000  45219.410156  44850.109375  44909.140625   
2019-04-11  44872.531250  44966.500000  44373.488281  44580.058594   
2019-04-12  44767.671875  44888.699219  44534.378906  44686.058594   
2019-04-15  44649.738281  44900.929688  44347.531250  44625.781250   

               Adj Close     Volume  
Date                                 
2019-04-09  45151.628906  191575000  
2019-04-10  44909.140625  145314900  
2019-04-11  44580.058594  109090000  
2019-04-12  44686.058594  143662400  
2019-04-15  44625.781250  108627100

In [48]:
# 5. Calculate the growth for both indices:
sp500_growth = (df_sp500_bz['Close'].iloc[-1] / df_sp500_bz['Close'].iloc[0]) - 1
ipc_mexico_growth = (df_ipc_mexico_bz['Close'].iloc[-1] / df_ipc_mexico_bz['Close'].iloc[0]) - 1

print(f"sp500 Growth: {sp500_growth}")
print(f"ipc mexico Growth: {ipc_mexico_growth}")


sp500 Growth: 0.8075151917783085
ipc mexico Growth: 0.2843377484045295


In [49]:
sp500_growth_v2 = (df_sp500_bz['Close'].pct_change() + 1).prod() - 1
ipc_mexico_growth_v2 = (df_ipc_mexico_bz['Close'].pct_change() + 1).prod() - 1

print(f"sp500 Growth: {sp500_growth_v2}")
print(f"ipc mexico Growth: {ipc_mexico_growth_v2}")

sp500 Growth: 0.8075151917783014
ipc mexico Growth: 0.2843377484045313


In [50]:
# 6. Compare the growth and select the higher growing index:
if sp500_growth > ipc_mexico_growth:
    higher_growth_index = 'S&P 500'
    higher_growth = sp500_growth
else:
    higher_growth_index = 'IPC Mexico'
    higher_growth = ipc_mexico_growth

print(f"The higher growing index is {higher_growth_index} with a growth of {round(higher_growth * 100)}%")


The higher growing index is S&P 500 with a growth of 81%


---
### Question 4. [Stocks OHLCV] 52-weeks range ratio (2023) for the selected stocks


**Find the largest range ratio [=(max-min)/max] of Adj.Close prices in 2023**


Download the 2023 daily OHLCV data on Yahoo Finance for top6 stocks on earnings (https://companiesmarketcap.com/most-profitable-companies/): 2222.SR,BRK-B, AAPL, MSFT, GOOG, JPM.

Here is the example data you should see in Pandas for "2222.SR": https://finance.yahoo.com/quote/2222.SR/history

Calculate maximum-minimim "Adj.Close" price for each stock and divide it by the maximum "Adj.Close" value.
Round the result to two decimal places (e.g. 0.1575 will be 0.16)

(Additional: why this may be important for your research?)


In [54]:
start_date = '2023-01-01'
end_date = '2023-12-31'

# List of stock symbols
top6_stocks = ['2222.SR', 'BRK-B', 'AAPL', 'MSFT', 'GOOG', 'JPM']

In [55]:
# Create an empty dictionary to store the dataframes
top6_stock_data = {}

# Download data for each stock
for stock in top6_stocks:
    top6_stock_data[stock] = pdr.get_data_yahoo(stock, start=start_date, end=end_date)


[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed


[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed


In [60]:
# Calculate maximum-minimum "Adj Close" price for each stock
results = {}
for stock, data in top6_stock_data.items():
    max_price = data['Adj Close'].max()
    min_price = data['Adj Close'].min()
    max_min_diff = max_price - min_price
    normalized_diff = round(max_min_diff / max_price, 2)
    results[stock] = normalized_diff

# Print the results
for stock, normalized_diff in results.items():
    print(f"{stock}: {normalized_diff}")

2222.SR: 0.21
BRK-B: 0.21
AAPL: 0.37
MSFT: 0.42
GOOG: 0.39
JPM: 0.28


In [61]:
# Find the stock with the largest range ratio
largest_range_ratio_stock = max(results, key=results.get)
largest_range_ratio = results[largest_range_ratio_stock]

print(f"The stock with the largest range ratio is {largest_range_ratio_stock} with a ratio of {largest_range_ratio}")

The stock with the largest range ratio is MSFT with a ratio of 0.42


In [58]:
# Calculate 52-week range ratio for each stock
results = {}
for stock, data in top6_stock_data.items():
    low_52week = data['Low'].min()
    high_52week = data['High'].max()
    range_ratio = round(low_52week / high_52week, 2)
    results[stock] = range_ratio

# Print the results
for stock, range_ratio in results.items():
    print(f"{stock}: {range_ratio}")

2222.SR: 0.8
BRK-B: 0.78
AAPL: 0.62
MSFT: 0.57
GOOG: 0.59
JPM: 0.72


### Question 5. [Stocks] Dividend Yield
**Find the largest dividend yield for the same set of stocks**

Use the same list of companies (2222.SR,BRK-B, AAPL, MSFT, GOOG, JPM) and download all dividends paid in 2023.
You can use `get_actions()` method or `.dividends` field in yfinance library (https://github.com/ranaroussi/yfinance?tab=readme-ov-file#quick-start)

Sum up all dividends paid in 2023 per company and divide each value by the closing price (Adj.Close) at the last trading day of the year.

Find the maximum value in % and round to 1 digit after the decimal point. (E.g., if you obtained $1.25 dividends paid and the end year stock price is $100, the dividend yield is 1.25% -- and your answer should be equal to 1.3)

In [19]:
import yfinance as yf

start_date = '2023-01-01'
end_date = '2023-12-31'

# List of stock symbols
top6_stocks = ['2222.SR', 'BRK-B', 'AAPL', 'MSFT', 'GOOG', 'JPM']
# top6_stocks = ['2222.SR']

In [20]:
# Create an empty dictionary to store the dataframes
top6_stock_actions_data = {}
top6_stock_history_data = {}
top6_stock_dividend_yields = {}

# Download data for each stock
for stock in top6_stocks:
    stock_ticker_instance = yf.Ticker(stock)
    actions_data = stock_ticker_instance.actions
    top6_stock_actions_data[stock] = stock_ticker_instance.actions

    # Filter actions_data for 2023
    actions_data_2023 = actions_data.loc[actions_data.index.year == 2023]

    # Sum up the dividends for 2023
    total_dividends = actions_data_2023['Dividends'].sum()

    # Fetch the closing price on the last trading day of 2023
    history_data = stock_ticker_instance.history(start=start_date, end=end_date)
    top6_stock_history_data[stock] = history_data
    closing_price = history_data['Close'].iloc[-1]

    # Calculate the dividend yield percentage
    dividend_yield = (total_dividends / closing_price) * 100

    # Store the dividend yield in the dictionary
    top6_stock_dividend_yields[stock] = dividend_yield

In [21]:
for stock, dividend_yield in top6_stock_dividend_yields.items():
    print(f"{stock}: {dividend_yield}")   

2222.SR: 2.770149819503256
BRK-B: 0.0
AAPL: 0.4940592304162832
MSFT: 0.7433143940964608
GOOG: 0.0
JPM: 2.409608190645897


In [22]:
# Find the maximum dividend yield and round to 1 decimal place
max_dividend_yield = round(max(top6_stock_dividend_yields.values()), 1)

print(f"The maximum dividend yield among the top 6 stocks in 2023 is: {max_dividend_yield}%")

The maximum dividend yield among the top 6 stocks in 2023 is: 2.8%


In [18]:
for stock, history_data in top6_stock_history_data.items():
    print(f"{stock}: {history_data.iloc[0]}")
    print(f"{stock}: {history_data.iloc[-1]}")

2222.SR: Open            2.825607e+01
High            2.847579e+01
Low             2.821213e+01
Close           2.847579e+01
Volume          1.569087e+06
Dividends       0.000000e+00
Stock Splits    0.000000e+00
Name: 2023-01-01 00:00:00+03:00, dtype: float64
2222.SR: Open            3.297727e+01
High            3.307675e+01
Low             3.277831e+01
Close           3.287779e+01
Volume          1.272045e+07
Dividends       0.000000e+00
Stock Splits    0.000000e+00
Name: 2023-12-28 00:00:00+03:00, dtype: float64


In [6]:
for stock, actions_data in top6_stock_actions_data.items():
    print(f"{stock}: {actions_data.tail(8)}")

2222.SR:                            Dividends  Stock Splits
Date                                              
2022-08-22 00:00:00+03:00   0.290727           0.0
2022-11-09 00:00:00+03:00   0.290727           0.0
2023-03-15 00:00:00+03:00   0.302364           0.0
2023-05-09 00:00:00+03:00   0.000000           1.1
2023-05-17 00:00:00+03:00   0.302400           0.0
2023-09-11 00:00:00+03:00   0.153000           0.0
2023-11-15 00:00:00+03:00   0.153000           0.0
2024-03-14 00:00:00+03:00   0.167000           0.0


---
### Question 6. [Exploratory] Investigate new metrics

**Free text answer**

Download and explore a few additional metrics or time series that might be valuable for your project and write down why (briefly).

---
### Question 7. [Exploratory] Time-driven strategy description around earnings releases

**Free text answer**

Explore earning dates for the whole month of April - e.g. using YahooFinance earnings calendar (https://finance.yahoo.com/calendar/earnings?from=2024-04-21&to=2024-04-27&day=2024-04-23). Compare with the previous closed earnings (e.g., recent dates with full data https://finance.yahoo.com/calendar/earnings?from=2024-04-07&to=2024-04-13&day=2024-04-08). 

Describe an analytical strategy/idea (you're not required to implement it) to select a subset companies of interest based on the future events data.